### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by CvX**
[WIP] There are still a couple of TODOs to be resolved.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Can you mark properties that are meant as read-only with `readonly` and if all are readonly, you could use the `Readonly` helper.

https://basarat.gitbooks.io/typescript/docs/types/readonly.html

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+ */
+export const chromeVersion: string;
+
+type RequireAtLeastOne<T, Keys extends keyof T = keyof T> =
+	Pick<T, Exclude<keyof T, Keys>>
+	& {
+		[K in Keys]-?: Required<Pick<T, K>>
+	}[Keys]
+
+interface _Choices<M, W, L, D> {
```
Comment (sindresorhus) [inline index.d.ts]: Use descriptive type names if more than one, for readability.

`M` => `Macos`
`W` => `Windows`
etc

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+export function activeWindow(): Electron.BrowserWindow;
+
+/**
+ * Run some JavaScript in the active or given window.
+ *
+ * @param code - TODO
+ * @param window - Default: Current window
+ * @returns A promise for the result of the executed code or a rejected promise if the result is a rejected promise.
+ */
+export function runJS(code: string, window?: Electron.BrowserWindow): Promise<unknown>;
```
Comment (sindresorhus) [inline index.d.ts]: Use default parameter for `window`.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+ * import {openNewGitHubIssue} from 'electron-util';
+ *
+ * openNewGitHubIssue({
+ * 	user: 'sindresorhus',
+ * 	repo: 'playground',
+ * 	body: 'Hello'
+ * });
+ */
+// TODO options type?
+export function openNewGitHubIssue(options: unknown): void;
```
Comment (sindresorhus) [inline index.d.ts]: Maybe add it to https://github.com/sindresorhus/new-github-issue-url Then you could just import the type here.

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+export interface DarkMode {
+	/**
+	 * Whether the macOS dark mode is enabled.
+	 * On Windows and Linux, it's `false`.
+	 */
+	isEnabled: boolean;
+
+	/**
+	 * The `callback` function is called when the macOS dark mode is toggled.
+	 * Returns a function, that when called, unsubscribes the listener.
```
Comment (sindresorhus) [inline index.d.ts]: This is not reflected in the type definition.

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+ * @returns A `boolean` of whether it's the first time your app is launched.
+ */
+export function isFirstAppLaunch(): boolean;
+
+export interface DarkMode {
+	/**
+	 * Whether the macOS dark mode is enabled.
+	 * On Windows and Linux, it's `false`.
+	 */
+	isEnabled: boolean;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	readonly isEnabled: boolean;
```

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+ * });
+ */
+// TODO options type?
+export function openNewGitHubIssue(options: unknown): void;
+
+export interface OpenUrlMenuItemOptions extends Electron.MenuItemConstructorOptions {
+	/**
+	 * TODO
+	 */
+	url: string;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	url?: string;
```

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+export const api: Electron.AllElectron | Electron.Remote;
+
+/**
+ * Check for various things.
+ */
+export const is: {
+	/**
+	 * Running on macOS.
+	 */
+	macos: boolean,
```
Comment (sindresorhus) [inline index.d.ts]: Use `readonly` on all of these.

File: index.d.ts (hunk 6 above)
Comment (CvX) [reply index.d.ts]: Here the `url` option is required: https://github.com/sindresorhus/electron-util/blob/e2ea4fd/index.js#L221-L223

File: index.d.ts (hunk 2 above)
Comment (CvX) [reply index.d.ts]: I'm not sure what do you mean. It's marked as an optional parameter, and its default value is documented?

File: index.d.ts (hunk 3 above)
Comment (CvX) [reply index.d.ts]: Done: https://github.com/sindresorhus/new-github-issue-url/pull/2

File: index.d.ts (hunk 2 above)
Comment (CvX) [reply index.d.ts]: Okay, so based on https://github.com/sindresorhus/electron-debug/pull/68#issuecomment-458140940 we're not changing this…

**Comment by sindresorhus** [comment]
Wow. Good catch. Let's change them to return `MenuItemConstructorOptions`.

**Comment by sindresorhus** [comment]
I'm gonna open an issue tomorrow about Electron supporting `MenuItem` too in the template builder, but that will probably take a long time, if ever.

**Comment by CvX** [comment]
> Wow. Good catch. Let's change them to return `MenuItemConstructorOptions`.

That's TypeScript at work! 😄I've been testing these type definitions inside Caprine's codebase, so it got flagged instantly. 🙂

**Comment by sindresorhus** [comment]
Woah. This was a massive undertaking and you did an amazing job at it! 🎉

**Comment by sindresorhus** [comment]
Published: https://github.com/sindresorhus/electron-util/releases/tag/v0.11.0

**Comment by sindresorhus** [comment]
> I'm gonna open an issue tomorrow about Electron supporting MenuItem too

Done: https://github.com/electron/electron/issues/16645

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+ */
+export const chromeVersion: string;
+
+type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
+
+type RequireAtLeastOne<T, Keys extends keyof T = keyof T> =
+	Omit<T, Keys>
+	& {
+		[K in Keys]-?: Required<Pick<T, K>>
+	}[Keys]
```
Comment (sindresorhus) [inline index.d.ts]: @CvX Would you be willing to submit this to https://github.com/sindresorhus/type-fest ? I'm trying to consolidate the types I use so they can be properly documented and tested in one place.

**Pull request merged**